OMCompiler/Compiler/SimCode/SimCodeUtil.mo

```diff
@@ -9263,21 +9298,114 @@ end dlowvarToSimvar;
+  initial_ := match(variability, causality)
+
+    // variability=CONSTANT Vs causality = output and local
+    case (SimCodeVar.CONSTANT(), SimCodeVar.OUTPUT()) then SimCodeVar.EXACT();
+    case (SimCodeVar.CONSTANT(), SimCodeVar.LOCAL()) then SimCodeVar.EXACT();
+
+    // variability=FIXED Vs causality = parameter, calculatedParameter and local
+    case (SimCodeVar.FIXED(), SimCodeVar.PARAMETER()) then SimCodeVar.EXACT();
+    case (SimCodeVar.FIXED(), SimCodeVar.CALCULATED_PARAMETER()) then getInitialAttributeHelperForParameters(dlow, isFixed, iterationVars);
+    case (SimCodeVar.FIXED(), SimCodeVar.LOCAL()) then getInitialAttributeHelperForParameters(dlow, isFixed, iterationVars);
+
+    // TODO varkind= tunable Vs causality = parameter, calculatedParameter and local, same like PARAM()
+
+    // variability=DISCRETE Vs causality = output and local
+    case (SimCodeVar.DISCRETE(), SimCodeVar.OUTPUT()) then getInitialAttributeHelper(dlow, isFixed, iterationVars);
+    case (SimCodeVar.DISCRETE(), SimCodeVar.LOCAL()) then getInitialAttributeHelper(dlow, isFixed, iterationVars);
+
+    // varkind=Continuous Vs causality = output and local
+    case (SimCodeVar.CONTINUOUS(), SimCodeVar.OUTPUT()) then getInitialAttributeHelper(dlow, isFixed, iterationVars);
+    case (SimCodeVar.CONTINUOUS(), SimCodeVar.LOCAL()) then getInitialAttributeHelper(dlow, isFixed, iterationVars);
+
+    // cases to handle invalid combination where initial is not allowed
+    case (_, _) then SimCodeVar.NONE_INITIAL();
+  end match;
+end setInitialAttribute;
+
+protected function getInitialAttributeHelperForParameters
+  "function which returns the initial attribute of a variable
+   with Variablity = parameter"
+  input BackendDAE.Var var;
+  input Boolean isFixed;
+  input list<DAE.ComponentRef> iterationVars "list of iterationvars from InitializationDAE";
+  output SimCodeVar.Initial initial_;
+algorithm
+  if not isFixed and ComponentReference.crefInLst(var.varName, iterationVars) then
+    initial_ := SimCodeVar.APPROX();
+  else
+    initial_ := SimCodeVar.CALCULATED(); // default value
+  end if;
+end getInitialAttributeHelperForParameters;
+
+protected function getInitialAttributeHelper
+  "function which returns the initial attribute of a variable
+   with Variablity = Continuous or Discrete"
+  input BackendDAE.Var var;
+  input Boolean isFixed;
+  input list<DAE.ComponentRef> iterationVars "list of iterationvars from InitializationDAE";
+  output SimCodeVar.Initial initial_;
+algorithm
+  //print("\n getInitialAttributeHelper :" + anyString(iterationVars) + "\n");
+
+  if isFixed then  // Variables with Constant Bindings
+    initial_ := SimCodeVar.EXACT();
+  elseif ComponentReference.crefInLst(var.varName, iterationVars) then  // Check Variable is an iterationVar
+    initial_ := SimCodeVar.APPROX();
+  else
+    initial_ := SimCodeVar.CALCULATED(); // default value
+  end if;
+end getInitialAttributeHelper;
+
+protected function setVariabilityAttribute
```

Comment:
This doesn't set the initial attribute, but it returns it and should therefore rather be called `getVariabilityAttribute`.